Util/src/main/java/io/deephaven/util/datastructures/intrusive/IntrusiveSoftLRU.java

```diff
@@ -84,18 +150,76 @@ public synchronized void touch(@NotNull T t) {
         }
     }
 
-    public synchronized void clear() {
-        Arrays.fill(softReferences, getNull());
+    /**
+     * Simply add the item to the array if it is new, potentially resizing if needed and switching to LRU mode.
+     *
+     * @param itemToTouch the item to touch
+     */
+    private void touchSimple(@NotNull final T itemToTouch) {
+        final int slot = adapter.getSlot(itemToTouch);
+        final SoftReference<T> owner = adapter.getOwner(itemToTouch);
+        if (slot == -1 || softReferences[slot] != owner) {
+            // We'll have to add it, we haven't seen it before. First, do we need to resize?
+            if (size < softReferences.length) {
+                // No, cool. Just add it.
+                final int slotForItem = size++;
+                adapter.setSlot(itemToTouch, slotForItem);
+                softReferences[slotForItem] = owner;
+            } else {
+                // Yep, resize, copy, and potentially switch to LRU mode.
+                doResize();
+                touch(itemToTouch);
+            }
+        }
+    }
+
+    /**
+     * Resize the cache by doubling, stopping at the maximum capacity. If maximum capacity is reached, then switch the
+     * structure into LRU mode.
+     *
+     */
+    private void doResize() {
+        // Time to reallocate or change internal structures
+        final int newCapacity = Math.min(softReferences.length << 1, maxCapacity);
+        final SoftReference<T>[] oldRefs = softReferences;
+        // noinspection unchecked
+        softReferences = new SoftReference[newCapacity];
+
+        // We have grown too large, switch to LRU mode.
+        if (newCapacity == maxCapacity) {
+            initializeLinks();
+            for (final SoftReference<T> reference : oldRefs) {
+                final T maybeItem = reference.get();
+                if (maybeItem != null) {
+                    // Clear the slot since we're changing ownership.
+                    adapter.setSlot(maybeItem, -1);
+                    touchLRU(maybeItem);
+                }
+            }
+        } else {
+            this.size = 0;
+            for (final SoftReference<T> reference : oldRefs) {
+                final T maybeItem = reference.get();
+                if (maybeItem != null) {
+                    final int slotForItem = size++;
+                    adapter.setSlot(maybeItem, slotForItem);
+                    softReferences[slotForItem] = reference;
+                }
+            }
+        }
```

Comment:
This entire block could be safely reduced to an array copy and a conditional initializeLinks. *Maybe* with an additional step to replace cleared references that aren't the null reference with the null reference. That would more accurately reflect the assumption that nothing in the `softReferences` array is ever `null`.
```suggestion
        final int oldCapacity = softReferences.length;
        final int newCapacity = Math.min(oldCapacity << 1, maxCapacity);
        softReferences = Arrays.copyOf(softReferences, newCapacity);
        Arrays.fill(softReferences, oldCapacity, newCapacity, getNull());
        if (newCapacity == maxCapacity) {
            initializeLinks();
        }
```

Reply:
Accepted -- I added a step before so that we can still do compaction early and maybe get lucky and not expand